File: SourceCode/Java_Generic/ChalmersGU/BaseAPI.java

// The following interfaces are not used in the implementations.
// Instead we use the corresponding standard Java API interfaces.
//
// To not get into trouble, we comment out everything in this file.

/*

/* *** ODSATag: ComparableADT *** */ /*
// Note: This is the same as the java.lang.Comparable interface
interface Comparable<E> {
    int compareTo(E other);  // Returns <0  if  this < other,
                             //      or  0  if  this == other,
                             //      or >0  if  this > other.
}
/* *** ODSAendTag: ComparableADT *** */


/* *** ODSATag: ComparatorADT *** */ /*
// Note: This is a subset of the java.util.Comparator interface
interface Comparator<E> {
    int compare(E one, E other);  // Returns <0  if  one < other,
                                  //      or  0  if  one == other,
                                  //      or >0  if  one > other.
}
/* *** ODSAendTag: ComparatorADT *** */


/* *** ODSATag: IteratorADT *** */ /*
// Note: This is a subset of the java.util.Iterator interface
interface Iterator<E> {
    boolean hasNext();  // Returns true if the iterator is not empty.
    E next();           // Yields the next element in the iterator.
}

// Note: This is a subset of the java.lang.Iterable interface
interface Iterable<E> {
    Iterator<E> iterator();  // Returns a new iterator.
}
/* *** ODSAendTag: IteratorADT *** */ /*

*/

File: SourceCode/Python/ChalmersGU/BaseAPI.py

#/* *** ODSATag: ComparableADT *** */
# Note: by implementing these methods,
# one can use the standard Python comparison operators (==, !=, <, <=, >, >=).
class Comparable:
    def __eq__(self, other): """Test if self == other."""
    def __ne__(self, other): """Test if self != other."""
    def __lt__(self, other): """Test if self < other."""
    def __le__(self, other): """Test if self <= other."""
    def __gt__(self, other): """Test if self > other."""
    def __ge__(self, other): """Test if self >= other."""
#/* *** ODSAendTag: ComparableADT *** */


#/* *** ODSATag: ComparatorADT *** */
# Python doesn't use the same interface for comparators as Java does.
# Instead they advocate to use a *key extractor*, which is a function
# that converts an element to another element which then can be compared
# using the natural ordering.
#/* *** ODSAendTag: ComparatorADT *** */


#/* *** ODSATag: IteratorADT *** */
# Note: by implementing these methods,
# one can loop over the elements in a standard Python for-loop.
class Iterator:
    def __iter__(self): """Returns the iterator itself."""
    def __next__(self): """Returns the next item. Raises StopIteration if there are no more elements."""

class Iterable:
    def __iter__(self): """Returns a new iterator."""
#/* *** ODSAendTag: IteratorADT *** */

